lib/engine-universal/src/executable.rs

```diff
@@ -0,0 +1,158 @@
+    pub(crate) custom_sections: PrimaryMap<SectionIndex, CustomSection>,
+    pub(crate) custom_section_relocations: PrimaryMap<SectionIndex, Vec<Relocation>>,
+    // The section indices corresponding to the Dwarf debug info
+    pub(crate) debug: Option<Dwarf>,
+    // the Trampoline for Arm arch
+    pub(crate) trampolines: Option<TrampolinesSection>,
+    pub(crate) compile_info: CompileModuleInfo,
+    pub(crate) data_initializers: Vec<OwnedDataInitializer>,
+    pub(crate) cpu_features: u64,
+}
+
+impl UniversalExecutable {
+    /// Deserialize a Module from a slice.
+    /// The slice must have the following format:
+    /// RKYV serialization (any length) + POS (8 bytes)
+    ///
+    /// # Safety
+    ///
+    /// This method is unsafe since it deserializes data directly
+    /// from memory.
+    /// Right now we are not doing any extra work for validation, but
+    /// `rkyv` has an option to do bytecheck on the serialized data before
+    /// serializing (via `rkyv::check_archived_value`).
+    pub unsafe fn deserialize(metadata_slice: &[u8]) -> Result<Self, DeserializeError> {
+        let archived = Self::archive_from_slice(metadata_slice)?;
+        let mut deserializer = SharedDeserializeMap::new();
+        RkyvDeserialize::deserialize(archived, &mut deserializer)
+            .map_err(|e| DeserializeError::CorruptedBinary(format!("{:?}", e)))
+    }
+
+    /// # Safety
+    ///
+    /// This method is unsafe.
+    /// Please check `SerializableModule::deserialize` for more details.
+    pub unsafe fn archive_from_slice<'a>(
+        data: &'a [u8],
+    ) -> Result<&'a rkyv::Archived<UniversalExecutable>, DeserializeError> {
+        if !data.starts_with(&MAGIC_HEADER) {
+            return Err(DeserializeError::Incompatible(
+                "the provided bytes are not wasmer-universal".to_string(),
+            ));
+        } else if data.len() < MAGIC_HEADER.len() + 8 {
+            return Err(DeserializeError::Incompatible(
+                "invalid serialized data".into(),
+            ));
+        }
+        let (archive, position) = data.split_at(data.len() - 8);
+        let mut position_value = [0u8; 8];
+        position_value.copy_from_slice(position);
+        Ok(rkyv::archived_value::<UniversalExecutable>(
+            archive,
+            u64::from_le_bytes(position_value) as usize,
+        ))
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ExecutableSerializeError {
+    #[error("could not write universal executable header")]
+    WriteHeader(#[source] std::io::Error),
+    #[error("could not serialize the executable data")]
+    Executable(
+        #[source]
+        CompositeSerializerError<std::io::Error, AllocScratchError, SharedSerializeMapError>,
+    ),
+    #[error("could not write the position information")]
+    WritePosition(#[source] std::io::Error),
+    #[error("could not flush the writer")]
+    Flush(#[source] std::io::Error),
+}
```

Comment:
Puzzling to see io-error while we operate only on in-memory buffers, but I guess that's dictated by some existing API somewhere. Long term, it's relatively important to straighten this out, as error handling is a particularly troublesome source of bugs for us.

Reply:
The `io::Error` is gone now.